org.lflang/src/org/lflang/cli/CliBase.java

```diff
@@ -111,9 +145,39 @@ public void doExecute(Io io, String[] args) {
     /**
      * The entrypoint of Picocli applications - the first method called when 
      * CliBase, which implements the Runnable interface, is instantiated.
-     * Lfc and Lff have their own specific implementations for this method.
      */ 
-    public abstract void run();
+    public void run() {
+        // If args are given in a json file, store its contents in jsonString.
+        if (topLevelArg.jsonFile != null) {
+            try {
+                topLevelArg.jsonString = new String(Files.readAllBytes(
+                        io.getWd().resolve(topLevelArg.jsonFile)));
+            } catch (IOException e) {
+                reporter.printFatalErrorAndExit(
+                        "No such file: " + topLevelArg.jsonFile);
+            }
+        }
+        // If args are given in a json string, (1) unpack them into an args
+        // array, and (2) call cmd.execute on them, which assigns them to their
+        // correct instance variables, then (3) recurses into run().
```

Comment:
I personally avoid having inline comments in this level of detail because they tend to get out of sync with the implementation. Furthermore, they are not always clear to begin with. For example, where do we recurse into `run()`? I see that `doRun` is called, but it doesn't look recursive. The result is that I generally do not read inline comments.

This is a matter of opinion, and it isn't very important, but I am just noting it down here because this is a persistent problem and possible point of controversy that cumulatively can have consequences over time.